classes/services/PKPSubmissionService.php

```diff
@@ -0,0 +1,646 @@
+							$method = 'get' . ucfirst($param);
+						}
+						if (!empty($method)) {
+							$compiled[$param] = $submission->{$method}();
+						}
+						break;
+				}
+			}
+
+			$output[] = $compiled;
+		}
+
+		\HookRegistry::call('Submission::toArray::output', array(&$output, $params, $submissions));
+
+		return $output;
+	}
+
+	/**
+	 * Get details about a submission's stage(s)
+	 *
+	 * @param $submission Submission
+	 * @param $stageIds array|int|null One or more stages to retrieve.
+	 *  Default: null. Will return data on all app stages.
+	 * @return array {
+	 *  `id` int stage id
+	 *  `label` string translated stage name
+	 *  `queries` array [{
+	 *    `id` int query id
+	 *    `assocType` int
+	 *    `assocId` int
+	 *    `stageId` int
+	 *    `sequence` int
+	 *    `closed` bool
+	 *   }]
+	 *  `statusId` int stage status. note: on review stage, this refers to the
+	 *    status of the latest round.
+	 *  `status` string translated stage status name
+	 *  `files` array {
+	 *    `count` int number of files attached to stage. note: this only counts
+	 *      revision files.
+	 *   }
+	 */
+	public function toArrayStageDetails($submission, $stageIds = null) {
+
+		if (is_null($stageIds)) {
+			$stageIds = Application::getApplicationStages();
+		} elseif (is_int($stageIds)) {
+			$stageIds = array($stageIds);
+		}
+
+		$stages = array();
+		foreach ($stageIds as $stageId) {
+
+			import('lib.pkp.classes.workflow.WorkflowStageDAO');
+			$workflowStageDao = DAORegistry::getDAO('WorkflowStageDAO');
+			$stage = array(
+				'id' => (int) $stageId,
+				'label' => __($workflowStageDao->getTranslationKeyFromId($stageId)),
+				'isActiveStage' => $submission->getStageId() == $stageId,
+			);
+
+			// Discussions in this stage
+			$stage['queries'] = array();
+			$request = Application::getRequest();
+			import('lib.pkp.classes.query.QueryDAO');
+			$queryDao = DAORegistry::getDAO('QueryDAO');
+			$queries = $queryDao->getByAssoc(
+				ASSOC_TYPE_SUBMISSION,
+				$submission->getId(),
+				$stageId,
+				$request->getUser()->getId() // Current user restriction should prevent unauthorized access
+			)->toArray();
+
+			foreach ($queries as $query) {
+				$stage['queries'][] = array(
+					'id' => (int) $query->getId(),
+					'assocType' => (int) $query->getAssocType(),
+					'assocId' => (int) $query->getAssocId(),
+					'stageId' => $stageId,
+					'sequence' => (int) $query->getSequence(),
+					'closed' => (bool) $query->getIsClosed(),
+				);
+			}
+
+			// Stage-specific statuses
+			switch ($stageId) {
+
+				case WORKFLOW_STAGE_ID_SUBMISSION:
+					import('lib.pkp.classes.stageAssignment/StageAssignmentDAO');
+					$stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO');
+					$assignedEditors = $stageAssignmentDao->editorAssignedToStage($submission->getId(), $stageId);
+					if (!$assignedEditors) {
+						$stage['statusId'] = 1; // @todo this should be abstracted to a documented constant
```

Comment:
No, `STATUS_QUEUED` is used for the submission as a whole. This is a case where stage statuses would be useful. But we only have review round and review assignment statuses.